WIP FB8-89: RBR's idempotent mode to make slave crash safe

percona-ysorokin opened the pull request:
FB8-89: RBR's idempotent mode to make slave crash safe

Jira issue: https://jira.percona.com/browse/FB8-89

Reference Patch: https://github.com/facebook/mysql-5.6/commit/6b65ec5
Reference Patch: https://github.com/facebook/mysql-5.6/commit/9278491

8.0-specific changes.
* Similarly to 'memroot_unordered_set' added 'memroot_unordered_map' class
  template to 'include/map_helpers.h'
* 'binlog_recover()' now uses 'xid_to_gtid_container'
  ('memroot_unordered_map<my_xid, Gtid>') instead of plain-C 'HASH'.
* In 'THD::set_trans_pos()' we now use 'owned_gtid' instead of third input
  parameter.
* Partially cherry-picked 'worker_last_gtid'-related changes in 'Slave_worker'
  class from https://github.com/facebook/mysql-5.6/commit/b1b882b
* Partially cherry-picked 'last_gtid'-related changes in 'Relay_log_info'
  class from https://github.com/facebook/mysql-5.6/commit/b1b882b
* Removed 'mysql-test/suite/rpl_recovery/include/*' MTR include files from
  the original commit as they are identical to
  'mysql-test/suite/rpl/include/*' and are not used.
* Removed 'mysql-test/suite/rpl_recovery/extension/*' MTR include files from
  the original commit as they are identical to
  'mysql-test/suite/rpl/extension/*' and are not used.
* Removed 'innodb_use_native_aio = 0' from the
  'mysql-test/suite/rpl_recovery/rpl_1slave_base.cnf' and added
  'innodb_write_io_threads=2' / 'innodb_read_io_threads=0' instead as in 8.0.

---------- https://github.com/facebook/mysql-5.6/commit/6b65ec5 ----------
RBR's idempotent mode to make slave crash safe

Summary:
Currently to ensure crash safety in slave a DB->GTID map is maintained
which is updated after every commit. The map is required to identify holes
in the sequence of gtids caused by using MTS. In RBR we can make use of
idempotency of row events to handle holes instead. We only track the max gtid
encountered and after a crash we turn on idempotency for all transactions which
have gtid less than the max gtid. If any trx was already executed in this range
it's okay because we've turned on idempotent mode.
Closes #384

Originally Reviewed By: santoshbanda

Pulled By: abhinav04sharma

fbshipit-source-id: 4c6da14

---------- https://github.com/facebook/mysql-5.6/commit/9278491 ----------
Fix the gtid used during idempotent recovery

Summary:
Idempotent recovery relies on the fact that the max engine gtid is greater than the gtid being applied now.
However the gtid comparison is broken. max engine gtid was initialized by hashing into global_sid_map.
But, recovering gtid (i.e the gtid being currently applied)  was initialized using rli->recovery_sid_map.
Hence the gtid comparison is broken and is potentially dangerous.

This fix ensures that rli->recovery_max_engine_gtid is also initialized using rli->recovery_sid_map

Originally Reviewed By: abhinav04sharma

fbshipit-source-id: ecdb0f28f96

percona-ysorokin commented:
This is an initial version of the prerequisite commit porting. There are some parts which are still WIP.
The reviewer is expected to do high-level evaluation of the patch.
Code fragments marked with `// Yura: ...` comment are expected to be compared with the original 5.6 patch.
Although significant amount of work was done on improving MTR test cases stability - some of them still require additional work (at least changing `sync_slave_with_master` to `--source include/sync_slave_sql_with_master.inc`). Therefore, their quality can be ignored for now.

abhinav04sharma commented:
Hi, I don't think this is a perquisite for the idempotent recovery patch. I'd just port over the tests, the code should not be required.

percona-ysorokin commented:
@hermanlee @abhinav04sharma I have just updated this PR with my latest code changes. It looks pretty stable. 'rpl_recovery' MTR test suite runs clean with `--big-test --repeat=2`.
There are a few places in the code which I wasn't sure about and would like ask your advise - they are marked with `// Yura:` comment.

I would also like to do some more stylistic / c++11-izing code changes and some more MTR test case polishing but what I am sharing here is pretty close to the final version.

Your feedback is appreciated and thanks for all your previous suggestions and explanations.

percona-ysorokin commented on mysql-test/suite/rpl_recovery/rpl_1slave_base.cnf:
```diff
@@ -0,0 +1,57 @@
+log-bin=                    slave-bin
+relay-log=                  slave-relay-bin
+
+loose-innodb_write_io_threads=    2
```
@hermanlee  These are the options I took from 8.0 `mysql-test/suite/rpl/rpl_1slave_base.cnf`. Without them I occasionally was getting crashes in `AIO` destructor
```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/home/yura/addon/vs/facebook-build-8.0/Linux-Debug/runtime_output_directory/mys'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __pthread_kill (threadid=<optimized out>, signo=11) at ../sysdeps/unix/sysv/linux/pthread_kill.c:57
57	../sysdeps/unix/sysv/linux/pthread_kill.c: No such file or directory.
[Current thread is 1 (Thread 0x7f288bafc700 (LWP 35446))]
#0  __pthread_kill (threadid=<optimized out>, signo=11) at ../sysdeps/unix/sysv/linux/pthread_kill.c:57
#1  0x000055a3b9759780 in my_write_core (sig=11) at /mnt/hgfs/repos/reference/facebook-mysql/feature/mysys/stacktrace.cc:317
#2  0x000055a3b83f2e3a in handle_fatal_signal (sig=11) at /mnt/hgfs/repos/reference/facebook-mysql/feature/sql/signal_handler.cc:249
#3  <signal handler called>
#4  0x000055a3b9909520 in AIO::~AIO (this=0x7f28842e2578, __in_chrg=<optimized out>) at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/innobase/os/os0file.cc:6159
#5  0x000055a3b99126ec in ut_allocator<AIO>::destroy (this=0x7f288baf4a40, p=0x7f28842e2578) at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/innobase/include/ut0new.h:669
#6  0x000055a3b9911798 in ut_delete<AIO> (ptr=0x7f28842e2578) at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/innobase/include/ut0new.h:968
#7  0x000055a3b99093da in AIO::create (id=LATCH_ID_OS_AIO_IBUF_MUTEX, n=256, n_segments=1) at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/innobase/os/os0file.cc:6128
#8  0x000055a3b990983e in AIO::start (n_per_seg=256, n_readers=4, n_writers=4, n_slots_sync=100) at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/innobase/os/os0file.cc:6219
#9  0x000055a3b990a414 in os_aio_init (n_readers=4, n_writers=4, n_slots_sync=100) at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/innobase/os/os0file.cc:6469
#10 0x000055a3b9a1667e in srv_start (create_new_db=false, scan_directories="./;./;./") at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/innobase/srv/srv0start.cc:1787
#11 0x000055a3b97bb541 in innobase_init_files (dict_init_mode=DICT_INIT_CHECK_FILES, tablespaces=0x7f288bafb930) at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/innobase/handler/ha_innodb.cc:4627
#12 0x000055a3b97cad42 in innobase_ddse_dict_init (dict_init_mode=DICT_INIT_CHECK_FILES, version=80013, tables=0x7f288bafb910, tablespaces=0x7f288bafb930) at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/innobase/handler/ha_innodb.cc:11169
#13 0x000055a3b9683f28 in dd::bootstrap::DDSE_dict_init (thd=0x55a3bc214100, dict_init_mode=DICT_INIT_CHECK_FILES, version=80013) at /mnt/hgfs/repos/reference/facebook-mysql/feature/sql/dd/impl/bootstrapper.cc:2142
#14 0x000055a3b9674e31 in dd::upgrade_57::do_pre_checks_and_initialize_dd (thd=0x55a3bc214100) at /mnt/hgfs/repos/reference/facebook-mysql/feature/sql/dd/upgrade/upgrade.cc:1080
#15 0x000055a3b850e328 in bootstrap::handle_bootstrap (arg=0x55a3bc214100) at /mnt/hgfs/repos/reference/facebook-mysql/feature/sql/bootstrap.cc:347
#16 0x000055a3b9d4bb30 in pfs_spawn_thread (arg=0x55a3bc21f4b0) at /mnt/hgfs/repos/reference/facebook-mysql/feature/storage/perfschema/pfs.cc:2851
#17 0x00007f289ac6d6db in start_thread (arg=0x7f288bafc700) at pthread_create.c:463
#18 0x00007f2897d5688f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
I don't think that this crash is somehow related to my RBR idempotent recovery changes. Most probably it was already in `fb-mysql-8.0.13` branch. These new `rpl_recovery` tests just helped to reveal this problem easier.

percona-ysorokin replied:
It looks like if creation/initialization of AIO object array fails, it's not safe to destroy a partially created array at line
https://github.com/facebook/mysql-5.6/blob/b5d9d0426353162a1529dcf22406c23334203b16/storage/innobase/os/os0file.cc#L6128

percona-ysorokin replied:
I believe this is connected to `Merge aio page read requests` implementation
https://github.com/facebook/mysql-5.6/commit/92a2c008bc93e5e323eb0b599b76bcfae6534029

hermanlee replied:
I've asked @yizhang82 to help take a look at these changes. In the backtrace above, do you know the reason array->init() failed?

percona-ysorokin replied:
@hermanlee I believe this is connected to AIO resource exhaustion when `--parallel=N` is pretty big.
From the log file just before the crash
```
2019-07-23T13:39:06.719515Z 1 [Warning] [MY-012582] [InnoDB] io_setup() failed with EAGAIN. Will make 5 attempts before giving up.
2019-07-23T13:39:06.719610Z 1 [Warning] [MY-012583] [InnoDB] io_setup() attempt 1.
2019-07-23T13:39:07.219882Z 1 [Warning] [MY-012583] [InnoDB] io_setup() attempt 2.
2019-07-23T13:39:07.720587Z 1 [Warning] [MY-012583] [InnoDB] io_setup() attempt 3.
2019-07-23T13:39:08.221431Z 1 [Warning] [MY-012583] [InnoDB] io_setup() attempt 4.
2019-07-23T13:39:08.721867Z 1 [Warning] [MY-012583] [InnoDB] io_setup() attempt 5.
2019-07-23T13:39:09.222156Z 1 [ERROR] [MY-012584] [InnoDB] io_setup() failed with EAGAIN after 5 attempts.
2019-07-23T13:39:09.222181Z 1 [Note] [MY-012587] [InnoDB] You can disable Linux Native AIO by setting innodb_use_native_aio = 0 in my.cnf
```
I am pretty sure I saw an error like this even on upstream but it never caused a server crash.

percona-ysorokin replied:
I looks like `AIO::init_linux_native_aio()` which is called from `AIO::init()` may return with an error status before `m_pending` array is initialized at https://github.com/facebook/mysql-5.6/blob/b5d9d0426353162a1529dcf22406c23334203b16/storage/innobase/os/os0file.cc#L6078
This causes an assertion in the AIO destructor.

percona-ysorokin replied:
The issue can easily be reproduced with the following MTR command
```
./mysql-test/mtr --debug-server --force --retry=0 --parallel=12 --repeat=2 rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob rpl.rpl_row_img_idx_noblob
```

percona-ysorokin replied:
Created
https://jira.percona.com/browse/FB8-234
and
https://github.com/facebook/mysql-5.6/pull/1038

percona-ysorokin committed:
FB8-89: RBR's idempotent mode to make slave crash safe

Jira issue: https://jira.percona.com/browse/FB8-89

Reference Patch: https://github.com/facebook/mysql-5.6/commit/6b65ec5
Reference Patch: https://github.com/facebook/mysql-5.6/commit/9278491

8.0-specific changes.
* Similarly to 'memroot_unordered_set' added 'memroot_unordered_map' class
  template to 'include/map_helpers.h'
* 'binlog_recover()' now uses 'xid_to_gtid_container'
  ('memroot_unordered_map<my_xid, Gtid>') instead of plain-C 'HASH'.
* In 'THD::set_trans_pos()' we now use 'owned_gtid' instead of third input
  parameter.
* Partially cherry-picked 'worker_last_gtid'-related changes in 'Slave_worker'
  class from https://github.com/facebook/mysql-5.6/commit/b1b882b
* Partially cherry-picked 'last_gtid'-related changes in 'Relay_log_info'
  class from https://github.com/facebook/mysql-5.6/commit/b1b882b
* Removed 'mysql-test/suite/rpl_recovery/include/*' MTR include files from
  the original commit as they are identical to
  'mysql-test/suite/rpl/include/*' and are not used.
* Removed 'mysql-test/suite/rpl_recovery/extension/*' MTR include files from
  the original commit as they are identical to
  'mysql-test/suite/rpl/extension/*' and are not used.
* Removed 'innodb_use_native_aio = 0' from the
  'mysql-test/suite/rpl_recovery/rpl_1slave_base.cnf' and added
  'innodb_write_io_threads=2' / 'innodb_read_io_threads=2' instead as in 8.0.
* Updated and re-recorded 'binlog_nogtid.binlog_persist_only_variables' and
  'binlog_nogtid.binlog_persist_variables' MTR test cases because of the new
  system variable 'slave_use_idempotent_for_recovery'.

---------- https://github.com/facebook/mysql-5.6/commit/6b65ec5 ----------
RBR's idempotent mode to make slave crash safe

Summary:
Currently to ensure crash safety in slave a DB->GTID map is maintained
which is updated after every commit. The map is required to identify holes
in the sequence of gtids caused by using MTS. In RBR we can make use of
idempotency of row events to handle holes instead. We only track the max gtid
encountered and after a crash we turn on idempotency for all transactions which
have gtid less than the max gtid. If any trx was already executed in this range
it's okay because we've turned on idempotent mode.
Closes #384

Originally Reviewed By: santoshbanda

Pulled By: abhinav04sharma

fbshipit-source-id: 4c6da14

---------- https://github.com/facebook/mysql-5.6/commit/9278491 ----------
Fix the gtid used during idempotent recovery

Summary:
Idempotent recovery relies on the fact that the max engine gtid is greater than the gtid being applied now.
However the gtid comparison is broken. max engine gtid was initialized by hashing into global_sid_map.
But, recovering gtid (i.e the gtid being currently applied)  was initialized using rli->recovery_sid_map.
Hence the gtid comparison is broken and is potentially dangerous.

This fix ensures that rli->recovery_max_engine_gtid is also initialized using rli->recovery_sid_map

Originally Reviewed By: abhinav04sharma

fbshipit-source-id: ecdb0f28f96

percona-ysorokin commented:
Pushed a fix for `binlog_nogtid.binlog_persist_only_variables` and `binlog_nogtid.binlog_persist_variables` MTR test cases along with various c++11-style changes and grammar fixes.

percona-ysorokin commented:
@hermanlee I also noticed that for some unknown reason a new system variable `slave_use_idempotent_for_recovery` is strangely defined as `Sys_var_enum` with 2 predefined values `YES` and `NO`.
I am tempted to convert it into `Sys_var_bool` unless you know about potential compatibility issues which this change could cause.
